### ISSUE TITLE
4.0.11: insufficient cgroups driver initialization configuration options for root-less nested (k8s) containers

### ISSUE THREAD

**Issue by sitano** [lxc/lxc#4116]
# Required information

I execute root-less LXC containers (user namespace based with "host" networking) inside of the privileged (root+noseccomp+nolsm) Kubernetes Pods (`containerd` containers) .

 * Distribution: 
   * Host uses Debian `Debian GNU/Linux 10 (buster)` with the kernel `SMP Debian 4.19.132-1 (2020-07-24) x86_64`. 
   * Pods use `Ubuntu 18.04.5 LTS` with `containerd` as a CRI.
   * Pods use LXC version based on `4.0.11` ([15515f9a3](https://github.com/lxc/lxc/commit/15515f9a3)).  
 * The output of
   * `lxc-start --version`: 4.0.11 ([15515f9a3](https://github.com/lxc/lxc/commit/15515f9a3))
   * `systemd` mode and `stat -f /sys/fs/cgroup/systemd/`: `systemd.unified_cgroup_hierarchy=0` mode according to the https://github.com/systemd/systemd/blob/69c0807432fa4fbfbf507a53872664cd26715559/src/basic/cgroup-util.c#L1998 (legacy, unified_systemd_v232 = false;):
   
```
$ stat -f /sys/fs/cgroup/systemd/
  File: "/sys/fs/cgroup/systemd/"
    ID: 0        Namelen: 255     Type: cgroupfs 
```

   * `cat /proc/self/cgroup` - host uses `HYBRID` cgroup hierarchy (/sys/fs/cgroup v1 + unified/ v2 )

```
11:perf_event:/kubepods/burstable/pod/id
10:devices:/kubepods/burstable/pod/id
9:pids:/kubepods/burstable/pod/id
8:freezer:/kubepods/burstable/pod/id
7:blkio:/kubepods/burstable/pod/id
6:net_cls,net_prio:/kubepods/burstable/pod/id
5:rdma:/
4:memory:/kubepods/burstable/pod/id
3:cpuset:/kubepods/burstable/pod/id
2:cpu,cpuacct:/kubepods/burstable/pod/id
1:name=systemd:/kubepods/burstable/pod/id
0::/system.slice/containerd.service
```

   * `cat /proc/1/mounts` - pay attention to the fact that cgroups v2 are not mounted at `/sys/fs/cgroup/unified` 

```
overlay / overlay rw,relatime,lowerdir=/var/lib/containerd/.../work 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
...
sysfs /sys sysfs ro,nosuid,nodev,noexec,relatime 0 0
tmpfs /sys/fs/cgroup tmpfs rw,nosuid,nodev,noexec,relatime,mode=755 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,name=systemd 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/rdma cgroup rw,nosuid,nodev,noexec,relatime,rdma 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
...
```

# Issue description

Thanks to this patch https://github.com/lxc/lxc/pull/3891 by @brauner / https://github.com/lxc/lxc/commit/a0f38dbf0ba6778ced26b3d42c1a08eb4ca4689b by @stgraber from Jul 1, 2021 the LXC obtained an ability to handle funky cgroups layouts. However, this only partially solved the problem for LXC containers executed inside of the crafted cgroups namespaces.

K8s pods have their own cgroup namespaces such that the paths of the cgroup classes from `/proc/{1,self}/cgroup` (pod's procfs) never correspond to the mounted cgroup filesystems. The pod's mount at `/sys/fs/cgroup` sees only a subtree at which it was cut by the cloning of the cgroup namespace. Thus the LXC cgfsns driver can't locate the aforementioned controllers classes based on the info from `/proc/{1,self}/cgroup`. This must be solved by the aforementioned patch (https://github.com/lxc/lxc/commit/a0f38dbf0ba6778ced26b3d42c1a08eb4ca4689b) but it leads to another problem.

## TL; DR;

When all the cgroups paths belong to the "funky" cgroup layout LXC can't detect the cgroup hierarchy layout correctly and there are not options to restrict available controllers. 

## Problem 1 - no successful paths

According to the code at `cgfsng.c/__initialize_cgroups` uses controllers from the `/proc/{1,self}/cgroup` to determine the type of the cgroups hierarchy that the system must use by opening their paths that never would succeed in my environment and in any environment executed within a cloned cgroup namespace (and remounted proc). That leads to the situation when the `cgroup_hierarchy_add(..., type)` will never be called as well as any of the assignments to the `ops->cgroup_layout = CGROUP_LAYOUT_*;` thx `continue; // (funky cgroup layout?)` so that the LXC will end with a failure to detect the proper cgroups layout to proceed: 

> `cgroups/cgroup.c:cgroup_init - Unsupported cgroup layout`.

Let's imagine we have an option to say:

> lxc.cgroup.namespace = true

That overrides all the paths from `/proc/{1,self}/cgroup` and they begin to match with the mounts at /sys/fs/cgroup/...

## Problem 2 - cgroup layout restriction

Let's imagine we have solved the problem 1 with `lxc.cgroup.namespace = true` and all controllers have been resolved successfuly (even tho in my configuration it will be detected that the `unified` hierarchy is not mounted). 

Now we have LXC resolved both cgroups versions 1 and 2 so that both can be used for LXC containers - HYBRID mode. But our host container system (containerd) does not use cgroups v2 for restricting processes resources for Pods processes, it only uses cgroups v1: `2:cpu,cpuacct:/kubepods/burstable/pod/id`. cgroups v2 in its turn are only used by the host init system (systemd) (`0::/system.slice/containerd.service`). and more over, the class in which our `/proc/{1,self}/cgroup` is present is not delegated to us in a systemd sence - we are not allowed to use it for our own purposes and modify it.

Probably it will be detected by the `unified_hierarchy_delegated` that the `0::/system.slice/containerd.service` is not a delegated cgroup and will be skipped with a warning. But what we really wanted is to tell the LXC system that the cgroups v2 shall not be used at all. It would be good to say - only this type of cgroups must be used for this containers. Somethink like:

> lxc.cgroup.type = LEGACY

or 

> lxc.cgroup.restrict.type = LEGACY

## Problem 3 - available controllers set

There are controllers that we don't want to offer to the LXC system. For example `freezer`, `rdma`, `devices`, `systemd` from cgroup v1. There is no way to restrict it. Some of the controllers may be not delegated at all (as ie `rdma` in my case lacks `+w` perm). For example `runsc` restricts them in the following way: https://github.com/google/gvisor/blob/master/runsc/cgroup/cgroup.go#L47-L65=.

It could be something like:

> lxc.cgroup.controllers = memory,cpu,cpuset  

## Problem 4 - logs

In the logs we can see the following error traces:

```
(a) cgroups/cgfsng.c:__initialize_cgroups:3262 - No such file or directory - Current cgroup 10/pods/something/podf53e4581-54ac-4956-8ee4-c696f65fb524/1bad82fb2f23afadb8e3ab2b26ec6345916a276402b5cface26360d8f19e78dd does not exist (funky cgroup layout?)
(b) cgroups/cgfsng.c:legacy_hierarchy_delegated:3060 - Permission denied - Legacy hierarchy not writable, skipping
(c) cgroups/cgfsng.c:__initialize_cgroups:3169 - No such file or directory - Unified cgroup not mounted
```

The problem is that they are not telling what they mean, how awful things are and actually what it is all about exactly:

- 4.a - lacks the controller name under the test
- 4.b - also lacks the controller name and the path which LXC detected as non-writable
- 4.c - "not mounted" is not what is happening. what is happening actually is "No such file or directory" for some controller path. It needs the controller and the path which failed.

## Question 1

Am I missing something regarding Problem 1 to resolve it without a patch? (it is of course possible to simulate the paths present at `/proc/{1,self}/cgroup` by just creating them but that probably not a good solution - the LXC system must support this kind of  configuration flawlessly)

## Question 2

What do you think of adding any of aforementioned configuration options? Would you like to have a patch? For my setup the critical is only Problem 1.

## Question 3

Why LXC does not ignore `systemd` cgroup at all?

# Steps to reproduce

Repeat the environment mentioned in the distribution section. Execute privileged pod. Try to start root-less container.

# Information to attach

 - [lxc-start-cgfsns-part-error.log](https://github.com/lxc/lxc/files/8682166/lxc-start-cgfsns-part-error.log) container log (The <log> file from running `lxc-start -n <c> -l TRACE -o <logfile> `)

**Comment by sitano** [comment]
What do you think?

**Comment by sitano** [comment]
The workaround for this case to solve Problem 1:

```
$ cat /proc/1/cgroup | awk 'FS=":" { if ($2 ~ "blkio|cpu|memory|hugetlb|net|pids" && $3 ~ "/pods/") { system("mkdir -p /sys/fs/cgroup/" $2 $3 " && chown -R user:user /sys/fs/cgroup/" $2 "/pods") }}'
```

**Comment by sitano** [comment]
Problem 1 partially solved with the

```
commit 0b6b230e34062168e80f73090baae9737b679da6
Author: Christian Brauner <christian.brauner@ubuntu.com>
Date:   Thu Jan 13 18:48:15 2022 +0100

    attach: don't pointlessly call cgroup_init()
    
    We can let attach detect that it is running on a cgroup layout without
    writable cgroup hierarchies. In that case attach can finish early and
    doesn't need to run the heavy-handed cgroup parsing code.
    
    Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>
```

**Comment by sitano** [comment]
and f1c4a17e7 - (3 months ago) cgroups: log fd of newly created cgroup - Christian Brauner for logging
